File: plot_Bench_Test.py
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
import numpy as np

# ==============================================================================
# 1. 读取数据
# ==============================================================================
csv_file = "active_chassis_data.csv"
try:
    df = pd.read_csv(csv_file, skipinitialspace=True)
except FileNotFoundError:
    print(f"错误：找不到 {csv_file}。请先运行 C++ 程序生成数据。")
    exit()

# 提取物理量
# 注意：在 main.cpp 改完后，ActualHeight_FL 应该就是我们输入的正弦波
displacement_mm = df["ActualHeight_FL"] * 1000  # mm (行程)
force_air_N = df["AirForce_FL"]                 # N (空气弹簧力)
stiffness_kN_m = df["Stiffness_FL"] / 1000      # kN/m (瞬时刚度)
damping_Ns_m = df["AirDamping_FL"]              # Ns/m (热力学阻尼)

# ==============================================================================
# 2. 绘制四大特性图 (Bench Test Analysis)
# ==============================================================================
fig = plt.figure(figsize=(16, 12))
gs = gridspec.GridSpec(2, 2)
plt.suptitle('Single Chamber Air Spring: Bench Test Characteristics', fontsize=20, weight='bold')

# ------------------------------------------------------------------------------
# 图 1: 示功图 (Force vs Displacement) - 最重要的特性图
# 含义：展示刚度(斜率)和迟滞耗能(面积)
# ------------------------------------------------------------------------------
ax1 = fig.add_subplot(gs[0, 0])
# 使用散点图并用颜色表示过程（压缩/拉伸），方便看轨迹方向
sc = ax1.scatter(displacement_mm, force_air_N, c=np.linspace(0, 1, len(df)), cmap='jet', s=3, alpha=0.5)

ax1.set_title('(1) Hysteresis Loop (Force vs Displacement)', fontsize=14, fontweight='bold')
ax1.set_xlabel('Suspension Deflection (mm)\n[<-- Compression | Extension -->]', fontsize=12)
ax1.set_ylabel('Air Spring Force (N)', fontsize=12)
ax1.grid(True, linestyle='--', alpha=0.5)

# 添加说明
ax1.text(displacement_mm.mean(), force_air_N.mean(), "Area = Energy Loss\n(Thermodynamic Damping)",
         ha='center', va='center', bbox=dict(facecolor='white', alpha=0.8))

# ------------------------------------------------------------------------------
# 图 2: 刚度特性 (Stiffness vs Displacement)
# 含义：展示几何非线性 (Geometric Nonlinearity)
# ------------------------------------------------------------------------------
ax2 = fig.add_subplot(gs[0, 1])
ax2.plot(displacement_mm, stiffness_kN_m, 'r-', linewidth=2.0)

ax2.set_title('(2) Stiffness Characteristic (k1)', fontsize=14, fontweight='bold')
ax2.set_xlabel('Suspension Deflection (mm)', fontsize=12)
ax2.set_ylabel('Instant Stiffness (kN/m)', fontsize=12)
ax2.grid(True, linestyle='--', alpha=0.5)

# ------------------------------------------------------------------------------
# 图 3: 阻尼特性 (Damping vs Displacement)
# 含义：展示热力学阻尼随行程的变化
# ------------------------------------------------------------------------------
ax3 = fig.add_subplot(gs[1, 0])
ax3.plot(displacement_mm, damping_Ns_m, 'g-', linewidth=2.0)

ax3.set_title('(3) Thermodynamic Damping (c1)', fontsize=14, fontweight='bold')
ax3.set_xlabel('Suspension Deflection (mm)', fontsize=12)
ax3.set_ylabel('Damping Coefficient (Ns/m)', fontsize=12)
ax3.grid(True, linestyle='--', alpha=0.5)

# ------------------------------------------------------------------------------
# 图 4: 动态响应时序 (Time Series)
# 含义：检查输入输出是否正常
# ------------------------------------------------------------------------------
ax4 = fig.add_subplot(gs[1, 1])
time = df["Time"]
ax4.plot(time, displacement_mm, 'b--', label='Displacement (mm)')
ax4.set_ylabel('Disp (mm)', color='blue')
ax4.tick_params(axis='y', labelcolor='blue')

ax4_twin = ax4.twinx()
ax4_twin.plot(time, force_air_N, 'k-', label='Force (N)')
ax4_twin.set_ylabel('Force (N)', color='black')

ax4.set_title('(4) Dynamic Response (Time Domain)', fontsize=14, fontweight='bold')
ax4.set_xlabel('Time (s)', fontsize=12)
ax4.grid(True, linestyle='--', alpha=0.5)

plt.tight_layout()
plt.show()

File: plot_Bench_Test.m
% =========================================================================
% 单腔空气弹簧：台架测试力学特性验证 (单图独立生成版)
% =========================================================================
clc; clear; close all;

% 1. 读取 CSV 数据
data = readtable('active_chassis_data.csv');

% 提取完整时域数据 (用于图4，展示完整的动态跟随过程)
t = data.Time;
disp_mm = data.ActualHeight_FL * 1000;      % 悬架行程 [mm]
force_N = data.AirForce_FL;                 % 弹簧力 [N]

% 提取稳态周期数据 (用于图1、2、3，只取最后1秒，防止迟滞回线重叠)
idx = data.Time >= 4.0;
disp_mm_steady = data.ActualHeight_FL(idx) * 1000;
force_N_steady = data.AirForce_FL(idx);
stiffness_kNm_steady = data.Stiffness_FL(idx) / 1000;
damping_Nsm_steady = data.AirDamping_FL(idx);

% 2. 基础页面与字体设置
fs = 22;  % 字体大小
lw = 2.5; % 线宽
% 设置单张图片的默认弹出尺寸 [左下角x, 左下角y, 宽度, 高度]
fig_pos = [200, 200, 800, 600]; 

% =========================================================================
% 图 1: 迟滞回线 (力-位移特性)
% =========================================================================
figure(1);
set(gcf, 'position', fig_pos, 'color', 'white');
plot(disp_mm_steady, force_N_steady, 'Color', [0.850 0.325 0.098], 'LineWidth', lw); 
set(gca, 'FontSize', fs);
xlabel('悬架行程 [mm] [<-- 压缩 | 拉伸 -->]', 'fontsize', fs);
ylabel('空气弹簧力 [N]', 'fontsize', fs);
title(' 迟滞回线 (力-位移特性)', 'fontsize', fs);
grid on; 

% =========================================================================
% 图 2: 非线性刚度特性
% =========================================================================
figure(2);
set(gcf, 'position', fig_pos + [40, -40, 0, 0], 'color', 'white'); % 窗口稍微错开方便查看
plot(disp_mm_steady, stiffness_kNm_steady, 'r-', 'LineWidth', lw);
set(gca, 'FontSize', fs);
xlabel('悬架行程 [mm]', 'fontsize', fs);
ylabel('瞬时刚度 [kN/m]', 'fontsize', fs);
title(' 非线性刚度特性 (k_1)', 'fontsize', fs);
grid on; 

% =========================================================================
% 图 3: 热力学阻尼特性
% =========================================================================
figure(3);
set(gcf, 'position', fig_pos + [80, -80, 0, 0], 'color', 'white');
plot(disp_mm_steady, damping_Nsm_steady, 'g-', 'LineWidth', lw);
set(gca, 'FontSize', fs);
xlabel('悬架行程 [mm]', 'fontsize', fs);
ylabel('阻尼系数 [Ns/m]', 'fontsize', fs);
title(' 热力学阻尼特性 (c_1)', 'fontsize', fs);
grid on; 

% =========================================================================
% 图 4: 动态时域响应
% =========================================================================
figure(4);
set(gcf, 'position', fig_pos + [120, -120, 0, 0], 'color', 'white');
[hAx, hLine1, hLine2] = plotyy(t, disp_mm, t, force_N);

set(hLine1, 'LineStyle', '--', 'Color', 'b', 'LineWidth', 2);
set(hLine2, 'LineStyle', '-', 'Color', 'k', 'LineWidth', 2);

set(hAx(1), 'ycolor', 'b', 'FontSize', fs);
set(hAx(2), 'ycolor', 'k', 'FontSize', fs);

xlabel('时间 [s]', 'fontsize', fs);
ylabel(hAx(1), '位移 [mm]', 'fontsize', fs);
ylabel(hAx(2), '支撑力 [N]', 'fontsize', fs);
title(' 动态时域响应', 'fontsize', fs);

legend([hLine1, hLine2], {'位移 [mm]', '支撑力 [N]'}, 'fontsize', 18, 'Location', 'northeast', 'Box', 'off');
grid on;